### PR TITLE
ENH: Add kafka benchmarks

### PR DESCRIPTION
**Pull request by kalinon**
This adds functionality to bench mark kafka using the built in kafka-*-perf-test tools.

https://docs.cloudera.com/runtime/7.0.3/kafka-managing/topics/kafka-manage-cli-perf-test.html
https://engineering.linkedin.com/kafka/benchmarking-apache-kafka-2-million-writes-second-three-cheap-machines    

Here is an example config & output:

Config: 

```
    - name: "noreplication"
      # This is the number of instances we will fire up of the kafka-producer/kafka-consumer binaries
      threads: 2
      replication: 1
      partitions: 16
      recordSize: 100
      records: 60000000
      # These can be any official producer Kafka options: https://kafka.apache.org/documentation/#producerconfigs
      extraProducerOpts:
        - "buffer.memory=671088640"
      consumersOnly: false
      producersOnly: false
      timeout: 10000
```
Consumer Results:
```
➜  ~ k logs kafkaclient-sample-noreplication-consumer-bfkhm
start.time, end.time, data.consumed.in.MB, MB.sec, data.consumed.in.nMsg, nMsg.sec, rebalance.time.ms, fetch.time.ms, fetch.MB.sec, fetch.nMsg.sec
2020-02-18 15:17:33:018, 2020-02-18 15:18:26:152, 5722.0592, 107.6911, 60000139, 1129223.0775, 3099, 50035, 114.3611, 1199163.3656
➜  ~ k logs kafkaclient-sample-noreplication-consumer-pbh58
start.time, end.time, data.consumed.in.MB, MB.sec, data.consumed.in.nMsg, nMsg.sec, rebalance.time.ms, fetch.time.ms, fetch.MB.sec, fetch.nMsg.sec
2020-02-18 15:17:33:572, 2020-02-18 15:18:49:220, 5722.0810, 75.6409, 60000368, 793152.0728, 3098, 72550, 78.8709, 827020.9235
```

### CODE REVIEW

**Comment by kalinon** [comment]
This is dying for the same reason as #167 it looks like

File: controllers/kafkabench/consumer_job.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+	job := k8s.NewPerfJob(objectMeta, "kafkabench", cr.Spec.Image, cr.Spec.PodConfig)
+	job.Spec.Parallelism = &ts.Threads
+
+	// Add init job to sleep, this allows the producer to queue up messages
+	initContainer := corev1.Container{
+		Name:            "kafka-consumer-init",
+		Image:           cr.Spec.Image.Name,
+		ImagePullPolicy: corev1.PullPolicy(cr.Spec.Image.PullPolicy),
+		Command:         []string{"/bin/sleep", "40"},
```
Comment (otto-von-bivouac) [inline controllers/kafkabench/consumer_job.go:45]: It might be helpful to have this configurable.

File: controllers/kafkabench/consumer_job.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	timeout := "10000"
+	if ts.Timeout != nil {
+		timeout = fmt.Sprintf("%d", *ts.Timeout)
+	}
+
+	return []string{
+		"/usr/bin/kafka-consumer-perf-test",
+		"--broker-list", brokers,
+		"--messages", fmt.Sprintf("%d", ts.Records),
+		"--threads", "1",
```
Comment (otto-von-bivouac) [inline controllers/kafkabench/consumer_job.go:71]: Does it make sense to make this configurable?

File: pkg/k8s/job.go [hunk 3]
```diff
@@ -49,6 +49,23 @@ func NewPerfJob(objectMeta metav1.ObjectMeta, app string, imageSpec perfv1alpha1
 		}
 	}
 
+	if podConfig.PodScheduling.Affinity == nil {
```
Comment (otto-von-bivouac) [inline pkg/k8s/job.go:52]: I suppose this fix is somehow related to the Kafka enhancement. Could you please let me know why is this required?

**Comment by otto-von-bivouac** [comment]
I'll give this a try with Banzai Cloud's Kafka Operator. They might also enjoy this piece of work as they do plenty of benchmarkin on k8s.

File: pkg/k8s/job.go (hunk 3 above)
Comment (kalinon) [reply pkg/k8s/job.go:52]: I believe this was to prevent an API error that disallows you to partially define the affinities. This pre-populates the sub affinities with empty objects. It allowed me to check the PodAffinity like so:

```
	if job.Spec.Template.Spec.Affinity.PodAffinity == nil {
```

File: controllers/kafkabench/consumer_job.go (hunk 2 above)
Comment (kalinon) [reply controllers/kafkabench/consumer_job.go:71]: It is obfuscated via the api. You can pass the desired threads to `KafkaTestSpec` and instead of setting the threads on the CLI tool it sets it on the job spec as the value for `Parallelism`

```
	job := k8s.NewPerfJob(objectMeta, "kafkabench", cr.Spec.Image, cr.Spec.PodConfig)
	job.Spec.Parallelism = &ts.Threads
```

File: controllers/kafkabench/consumer_job.go (hunk 1 above)
Comment (kalinon) [reply controllers/kafkabench/consumer_job.go:45]: Yes, i dont particularly enjoy that being there. It probably should be replaced with an active check. I experimented with various values and 40 secs was about right. Theres a balance of too long or too short. If we wait too long, the producers finish with all their messages, too short and we listen before there are any.

Ill make this as a configurable. It makes sense.

Comment (kalinon) [reply controllers/kafkabench/consumer_job.go]: Done
